inspectors/view/qml/notation/fretdiagrams/FretCanvas.qml

```diff
@@ -0,0 +1,180 @@
+            ctx.textBaseline = "bottom";
+            var marker = (markerType === FretDiagramTypes.MARKER_CIRCLE) ? 'O' : 'X';
+            ctx.fillText(marker, stringDist * i, -fretDist * 0.1);
+        }
+
+        function drawDot(ctx, s, f, dotType) {
+            ctx.beginPath();
+            switch (dotType) {
+                case FretDiagramTypes.DOT_CROSS:
+                    ctx.beginPath();
+                    ctx.moveTo(s * stringDist - dotD * 0.5, fretDist * 0.5 + f * fretDist - dotD * 0.5);
+                    ctx.lineTo(s * stringDist + dotD * 0.5, fretDist * 0.5 + f * fretDist + dotD * 0.5);
+                    ctx.moveTo(s * stringDist + dotD * 0.5, fretDist * 0.5 + f * fretDist - dotD * 0.5);
+                    ctx.lineTo(s * stringDist - dotD * 0.5, fretDist * 0.5 + f * fretDist + dotD * 0.5);
+                    ctx.stroke();
+                    break;
+                case FretDiagramTypes.DOT_SQUARE:
+                    ctx.rect(s * stringDist - dotD * 0.5, fretDist * 0.5 + f * fretDist - dotD * 0.5, dotD, dotD);
+                    ctx.stroke();
+                    break;
+                case FretDiagramTypes.DOT_TRIANGLE:
+                    ctx.beginPath();
+                    ctx.moveTo(s * stringDist - dotD * 0.5, fretDist * 0.5 + f * fretDist + dotD * 0.5);
+                    ctx.lineTo(s * stringDist, fretDist * 0.5 + f * fretDist - dotD * 0.5);
+                    ctx.lineTo(s * stringDist + dotD * 0.5, fretDist * 0.5 + f * fretDist + dotD * 0.5);
+                    ctx.closePath();
+                    ctx.stroke();
+                    break;
+                case FretDiagramTypes.DOT_NORMAL:
+                default:
+                    ctx.ellipse(s * stringDist - dotD * 0.5, fretDist * 0.5 + f * fretDist - dotD * 0.5, dotD, dotD);
+                    ctx.stroke();
+                    ctx.fill();
+                    break;
+            }
+        }
+
+        function drawBarre(ctx, f, startString, endString) {
+            if (startString === -1)
+                return;
+            var x1 = stringDist * startString;
+            var x2 = (endString === -1) ? stringDist * (strings - 1)  : stringDist * endString;
+            var y  = fretDist * 0.5 + fretDist * f;
+            ctx.lineWidth = dotD * (model ? model.barreLineWidth() : 1.0);
+            ctx.lineCap = "round";
+            ctx.beginPath();
+            ctx.moveTo(x1, y);
+            ctx.lineTo(x2, y);
+            ctx.stroke();
+        }
+
+        onPaint: {
+            var ctx = getContext("2d");
+            ctx.reset();
+
+            ctx.strokeStyle = Qt.rgba(0, 0, 0, 1);
+            ctx.fillStyle = Qt.rgba(0, 0, 0, 1);
+            ctx.lineWidth = lineWidth1;
+
+            ctx.translate((width - diagramWidth) * 0.5, (height - diagramHeight) * 0.5);
+            ctx.save();
+
+            for (var s1 = 0; s1 < strings; s1++) {
+                var m = model ? model.marker(s1) : 0;
```

Comment:
I think it needs to be done differently. It is necessary to return the entire diagram object from the model and draw it, and not receive the diagram data in parts.
Like that: 
```
// C++
class Model
{
public:

    Q_INVOKABLE QVariantMap fretDiagram();
};

QVariantMap Model::fretDiagram()
{
    QVariantMap fd;
    QVariantList strings;

    QVariantMap string;
    string["marker"] = ...;
    strings << string;

    fd["strings"] = strings;
    
    QVariantMap style;
    style["barreLineWidth"] = ...;
    fd["style"] = style;

    return fd;
}

// Qml 

function drawFretDiagram(fd) {
    
    for (var s = 0; s < fd.strings.lenght; ++s) {
        var string = fd.strings[s];
        drawMarker(string.marker);
        ...
    }
} 

```
Put everything you need for drawing in the FretDiagram object. Also, it is possible to make fretDiagram a property (instead of Q_INVOKABLE) and the fretDiagramChanged signal has emitted if it changes.